teps/0045-guard-finally-tasks-with-whenexpressions.md

```diff
@@ -0,0 +1,303 @@
+when a certain `Task` in the `Pipeline` failed. To do this, one user has had to use a workaround using `Workspaces` that
+they describe [in this thread](https://tektoncd.slack.com/archives/CK3HBG7CM/p1603399989171300?thread_ts=1603376439.161500&cid=CK3HBG7CM).
+In addition, needing the workaround prevents the user from reusing the Slack catalog task as further described in [this issue](https://github.com/tektoncd/pipeline/issues/3438).
+
+We already guard `Tasks` using [`WhenExpressions`](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#guard-task-execution-using-whenexpressions),
+which efficiently evaluate the criteria of executing `Tasks`. We propose supporting using `WhenExpressions` to guard 
+the execution of `Finally Tasks` as well. 
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+The primary goal of this proposal is to allow guarding execution of `Finally Tasks` using `WhenExpressions`. A secondary
+goal is to improve the [reusability](https://github.com/tektoncd/community/blob/master/design-principles.md#reusability)
+of `Tasks` by improving the guarding of `Finally Tasks` at authoring time. 
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+This proposal is not enabling guarding `Finally Tasks` based on execution status of other `Finally Tasks`.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose supporting guarding `Finally Tasks` using `WhenExpressions`. The `WhenExpressions` can operate on static 
+inputs or variables such as `Parameters`, `Results` and `Execution Status`. If the `WhenExpressions` evaluate to `True`, 
+the `Finally Task` would be executed. If the `WhenExpressions` evaluate to `False`, the `Finally Task` would be skipped.
+
+Users would be able to solve for the example use case described in [Motivation](#motivation), where a user wants to send
+a Slack notification when a `Task` fails, as such:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: pipelinerun-
+spec:
+  pipelineSpec:
+    tasks:
+      - name: build
+        taskRef:
+          name: build
+      # […]
+    finally:
+      - name: notify-failure # executed only when build task fails
+        when:
+          - input: failure
+            operator: in
+            values:
+              - $(tasks.build.status)
```

Comment:
This looks little weird. can we change the when expression to switch input with values?

```
        when:
          - input: $(tasks.build.status)
            operator: in
            values: ["Failed"]
```

Reply:
done, thanks :)